subprojects/core/src/main/java/org/gradle/internal/buildtree/BuildTreeScopeServices.java

```diff
@@ -99,7 +100,7 @@ protected ExceptionAnalyser createExceptionAnalyser(LoggingConfiguration logging
         return exceptionAnalyser;
     }
 
-    protected DecompressionCacheFactory createDecompressionCacheFactory(BuildTreeScopedCache cacheFactory) {
-        return new DefaultDecompressionCacheFactory(cacheFactory);
+    protected DecompressionCacheFactory createDecompressionCacheFactory(ServiceRegistry serviceRegistry) {
+        return new DefaultDecompressionCacheFactory(() -> serviceRegistry.get(BuildTreeScopedCache.class));
```

Comment:
It seems a bit fragile to rely on the instantiation order of `BuildTreeScopedCache` to make this lazy. If the next service to come around does not consume `BuildTreeScopedCache` lazily, this will break. 

IMO the way forward here would be similar to below -- making `BuildTreeScopedCache` lazy as opposed to instantiating it later. That would subsequently mean making `ProjectCacheDir#getDir` and `BuildScopeCacheDir#getDir` also lazy. Though, this would avoid any worries of instantiation order causing us to use the wrong cache dir.

Reply:
We discussed this.  Reverting to injecting the `BTSC` and using a dumb supplier which just returns the injected instance avoids this issue.